### PR TITLE
fix(dag): enforce task-scoped step ids

### PR DESCRIPTION
**Pull request by AlexLiu190625**
## Background

DAG plans use LLM-generated step IDs and names. In multi-turn execution, especially during continuation or plan extension, the model can reuse an earlier step ID or emit the same ID more than once in a single plan response. That creates ambiguity for API and SDK consumers because trace events, step start/end events, and tool/LLM action events all rely on `step_id` to identify the DAG node they belong to.

Trace events already have unique event IDs, but those identify individual events, not stable DAG steps. The step ID still needs to be unique within a task so consumers can safely correlate all events for a step. Step names remain display labels and may repeat naturally.

## What changed

- Added task-scoped forbidden step ID collection before plan generation.
  - Existing IDs are collected from the current plan and execution history.
  - The planning prompts now include an explicit `FORBIDDEN_STEP_IDS` block.
  - Prompts clarify that step IDs must be unique, while step names do not need to be unique.

- Added hard validation for generated step IDs.
  - Empty IDs are rejected.
  - IDs already used in the same task are rejected.
  - Duplicate IDs within the same model response are rejected.
  - `_validate_plan()` now catches duplicate or empty IDs even when called directly.

- Kept the retry-first behavior.
  - Step ID validation failures are surfaced back into the retry prompt with the conflicting IDs and forbidden ID list.
  - The model gets a chance to regenerate a valid plan before any fallback rewrite is used.

- Added deterministic fallback rewriting after retries are exhausted.
  - Conflicting IDs are rewritten with stable suffixes, for example `analyze_2`.
  - Dependencies and conditional branch targets are rewritten to match the final generated IDs.
  - References to historical step IDs are preserved so new steps can still depend on previous steps.

- Tightened extension validation without revalidating old tools.
  - New extension steps are still checked against the current tool list.
  - The combined current plan plus new steps is checked for step ID uniqueness only, so historical steps are not rejected just because their original tools are not present in the current continuation context.

## Validation

- `uv run ruff check src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py tests/core/agent/pattern/dag_plan_execute/test_plan_validation_retry.py`
- `uv run ruff format src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py tests/core/agent/pattern/dag_plan_execute/test_plan_validation_retry.py`
- `uv run pytest tests/core/agent/pattern/dag_plan_execute/test_plan_generator.py tests/core/agent/pattern/dag_plan_execute/test_plan_validation_retry.py tests/core/agent/pattern/dag_plan_execute/test_conditional_branches.py`

Result: 37 passed, 1 skipped. The remaining warnings are third-party deprecation warnings from `deepdoc` imports.

### CODE REVIEW

File: src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py [hunk 1]
```diff
@@ -71,6 +72,260 @@ def __init__(
...
+        reserved_matches: List[Dict[str, Any]] = []
+        positions_by_id: Dict[str, List[int]] = {}
+
+        for index, step_data in enumerate(steps_data):
+            step_id = self._normalize_step_id_value(step_data.get("id"))
+            if not step_id:
+                missing_indices.append(index)
+                continue
+
+            step_data["id"] = step_id
```
Comment (rogercloud) [inline src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py:252]: This normalizes the step id but leaves dependency entries and conditional branch targets in their original form. If the model returns `id: " step1 "` and another step depends on `" step1 "`, the id becomes `step1` while the dependency still has spaces, so the later dependency cleanup treats it as invalid and removes it. That can make dependent steps run before their prerequisite. Please normalize dependencies and branch target values with the same `_normalize_step_id_value()` before validation/removal, and add a regression test for whitespace-padded references.

File: src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py [hunk 2]
```diff
@@ -71,6 +72,260 @@ def __init__(
...
+        steps_data: List[Dict[str, Any]],
+        forbidden_step_ids: set[str],
+    ) -> None:
+        """Reject empty, duplicate, or task-reserved step IDs in raw LLM output."""
+        missing_indices: List[int] = []
+        reserved_matches: List[Dict[str, Any]] = []
+        positions_by_id: Dict[str, List[int]] = {}
+
+        for index, step_data in enumerate(steps_data):
+            step_id = self._normalize_step_id_value(step_data.get("id"))
```
Comment (rogercloud) [inline src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py:247]: This missing-id check does not currently catch omitted IDs because `_parse_plan_response()` still injects `step_{i + 1}` whenever the `id` field is absent before validation runs. That contradicts the new contract that empty/missing IDs are rejected and surfaced through the retry prompt: a response with no `id` succeeds as `step_1` without retrying. Please preserve missing IDs through parsing, or mark them explicitly so `_validate_raw_step_ids()` can reject them.

File: src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py [hunk 3]
```diff
@@ -71,6 +72,260 @@ def __init__(
...
+
+    def _rewrite_step_reference(
+        self,
+        step_ref: Any,
+        forbidden_step_ids: set[str],
+        first_generated_id_by_original: Dict[str, str],
+    ) -> str:
+        """Rewrite dependency/branch references to the first generated final ID."""
+        normalized_ref = self._normalize_step_id_value(step_ref)
+        if normalized_ref in forbidden_step_ids:
```
Comment (rogercloud) [inline src/xagent/core/agent/pattern/dag_plan_execute/plan_generator.py:345]: When fallback rewrites a new step whose original id is forbidden, this branch keeps every reference to that original id pointing at the historical step. That is only safe when the reference truly meant the historical step. If the new response contains a conditional self-reference or another new step depends on the rewritten step, the reference is left as the forbidden id; in fresh-plan validation the branch target is then missing, and in extensions it can silently point at an old step instead of the newly rewritten one. Please distinguish references to rewritten new steps from intentional historical dependencies, or avoid fallback rewriting ambiguous reserved-id references and keep retrying/failing with a clear error.

**Review by rogercloud — CHANGES_REQUESTED** [review]
Thanks for tightening DAG step-id handling. I found a few edge cases where the new validation/rewrite path can still lose dependencies or fail to enforce the stated guarantees.

**Review by rogercloud — APPROVED** [review]
Reviewed the DAG plan-generator changes and ran the focused DAG plan-execute test package. No blocking issues found.

Validation:
- `uv run --group dev pytest tests/core/agent/pattern/dag_plan_execute` -> 63 passed, 1 skipped
- `uv run ruff check` on touched files -> passed
